File: frontend-vue/src/api/story.ts
import { http } from './http'

// 固定剧本 API：未传 story_id 时恢复既有 docs/story/07 第一章；
// story_id=prologue 时推进 docs/story/Prologue.md 的无序探班流程（docs/19）。
// 两者的后端权威游标均随会话/存档快照持久化。

export interface StoryLineNode {
  kind: 'line'
  speaker: string
  text: string
  emotion: string | null
  scene_id: string | null
}

export interface StoryOptionView {
  id: string
  label: string
}

export interface StoryChoiceNode {
  kind: 'choice'
  choice_id: string
  scene_id: string | null
  options: StoryOptionView[]
}

export interface StoryEndNode {
  kind: 'end'
  scene_id: string | null
}

export interface StoryChatNode {
  kind: 'chat'
  character_id: string
  scene_id: string | null
}

export type StoryNode = StoryLineNode | StoryChoiceNode | StoryEndNode | StoryChatNode

/** 场景演出指令（纯表现，后端权威；docs/17 演出接线） */
export interface StorySceneView {
  scene_id: string
  title: string
  presentation: {
    /** 场景入场脉冲播放的命名效果（SCREEN_GLITCH / SCREEN_SHAKE） */
    effects?: string[]
    /** GameBackground 光照滤镜（如 SC03 暗版 brightness 0.45） */
    lighting?: { background?: { brightness?: number } }
    /** 场景背景与权威在场角色（docs/19 序章）。 */
    background?: string
    characters?: Array<{
      character_id: string
      emotion: string
      slot?: string | null
      scale?: number
      offset_y?: number
    }>
  }
}

export interface StoryChapterOpening {
  chapter_label: string
  title: string
  /** 永远是章节首场景背景，不随当前恢复场景改变。 */
  background: string
}

export interface StoryView {
  session_id: string
  started: boolean
  finished: boolean
  node: StoryNode | null
  /** 当前节点所属场景（end 节点为 null） */
  scene: StorySceneView | null
  scene_changed: boolean
  chapter_opening: StoryChapterOpening
}

/** 当前展示节点（刷新/读档恢复；不移动游标）。 */
export async function fetchStoryCurrent(
  sessionId: string | null,
  storyId?: string,
): Promise<StoryView> {
  const { data } = await http.get<StoryView>('/story/current', {
    params: {
      ...(sessionId ? { session_id: sessionId } : {}),
      ...(storyId ? { story_id: storyId } : {}),
    },
  })
  return data
}

/** 「继续」：移动到下一节点（首次调用即开始故事）。 */
export async function storyAdvance(sessionId: string | null, storyId?: string): Promise<StoryView> {
  const { data } = await http.post<StoryView>('/story/advance', {
    session_id: sessionId,
    story_id: storyId ?? null,
  })
  return data
}

/** 提交一个 A/B/C 选项，返回该选项的第一句台词。 */
export async function storyChoose(
  sessionId: string,
  optionId: string,
  storyId?: string,
): Promise<StoryView> {
  const { data } = await http.post<StoryView>('/story/choose', {
    session_id: sessionId,
    option_id: optionId,
    story_id: storyId ?? null,
  })
  return data
}

File: backend/app/api/story.py
"""快速上线固定剧本 API（临时组件，对应 story_runtime.py）。

停用 AI 回复期间，前端只用这三个端点推进剧本：
- GET  /api/story/current?session_id=...   —— 刷新/读档后恢复当前展示节点
- POST /api/story/advance {session_id?, player_id?} —— 「继续」：下一节点
- POST /api/story/choose {session_id, option_id, player_id?} —— 选项 A/B/C

session_id 为空时 advance 由 orchestrator 铸造新会话（与 /api/chat 同约定）。
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field

from app.api.authz import bind_session, current_user_id, require_owned_session
from app.game.orchestrator import GameOrchestrator

router = APIRouter()


class StoryAdvanceRequest(BaseModel):
    session_id: str | None = None
    # docs/13 §15：匿名浏览器命名空间，供场景边界 AUTO 自动存档绑定。
    player_id: str | None = Field(default=None, max_length=64)


class StoryChooseRequest(BaseModel):
    session_id: str
    option_id: str = Field(min_length=1, max_length=64)
    player_id: str | None = Field(default=None, max_length=64)


class StoryNodeView(BaseModel):
    kind: str
    speaker: str | None = None
    text: str | None = None
    emotion: str | None = None
    choice_id: str | None = None
    scene_id: str | None = None
    options: list[dict] = []


class StoryView(BaseModel):
    session_id: str
    started: bool
    finished: bool
    node: StoryNodeView | None = None
    # 当前节点所属场景的标题与演出指令（纯表现数据；end 节点为 None）
    scene: dict | None = None
    scene_changed: bool = False


def get_orchestrator(request: Request) -> GameOrchestrator:
    return request.app.state.orchestrator


@router.get("/api/story/current", response_model=StoryView)
def story_current(
    request: Request,
    session_id: str | None = None,
    orchestrator: GameOrchestrator = Depends(get_orchestrator),
) -> StoryView:
    """当前展示节点。不移动游标；未知会话由 orchestrator 铸造新会话（游标
    未开始，前端随后用 advance 起步）。"""
    require_owned_session(request, session_id)
    try:
        view = orchestrator.story_current(session_id)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    bind_session(request, view["session_id"])
    return StoryView(**view)


@router.post("/api/story/advance", response_model=StoryView)
def story_advance(
    payload: StoryAdvanceRequest,
    request: Request,
    orchestrator: GameOrchestrator = Depends(get_orchestrator),
) -> StoryView:
    """「继续」：移动到下一节点。首次调用即开始故事。"""
    require_owned_session(request, payload.session_id)
    try:
        view = orchestrator.story_advance(
            payload.session_id, player_id=current_user_id(request, payload.player_id)
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    bind_session(request, view["session_id"])
    return StoryView(**view)


@router.post("/api/story/choose", response_model=StoryView)
def story_choose(
    payload: StoryChooseRequest,
    request: Request,
    orchestrator: GameOrchestrator = Depends(get_orchestrator),
) -> StoryView:
    """提交一个 A/B/C 选项，返回该选项的第一句台词。"""
    require_owned_session(request, payload.session_id)
    try:
        view = orchestrator.story_choose(
            payload.session_id,
            payload.option_id,
            player_id=current_user_id(request, payload.player_id),
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return StoryView(**view)
